From 16d506a9ebb4cbcb6d356f738bd80aff427148d8 Mon Sep 17 00:00:00 2001
From: Neha Bisht <quic_nbisht@quicinc.com>
Date: Wed, 10 Jul 2024 11:14:46 +0530
Subject: [PATCH] QSDK: wifi: ath12k: Fix to fetch correct primary_link_id for
 STA

Currently, for sta ahvif->primary_link_id is used to get arvif which is
only updated by debugfs. This results in null arvif dereferenced due to
incorrect primary_link in cases where sta is associated with primary
link other than 0.
Hence, instead of ahvif->primary_link, use ahsta->primary_link_id for
sta mode as well.
Also, in fse_add_flow and fse_delete_flow hardcoded link 0 is used,
instead use deflink/valid_links to get active link to fetch correct
arvif during ECM query.

fixes: I9dacaefcb (QSDK: wifi: ath12k: Add DS node id retrieval mechanism)
Ib3037ac47 (wifi: ath12k: Fix default link reference issue when DS enabled)

Patch dependency: 951-c-QSDK-wifi-ath12k-support-DS-for-SLO-and-MLO-AP-modes.patch
                  953-QSDK-wifi-ath12k-Add-DS-node-id-retrieval-mechanism.patch

Signed-off-by: Neha Bisht <quic_nbisht@quicinc.com>
Signed-off-by: Pavankumar Nandeshwar <quic_pnandesh@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/accel_cfg.c | 35 +++++++++++++++------
 drivers/net/wireless/ath/ath12k/ppe.c       | 14 ++++++---
 2 files changed, 35 insertions(+), 14 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/accel_cfg.c b/drivers/net/wireless/ath/ath12k/accel_cfg.c
index 368964a..d1cef6d 100644
--- a/drivers/net/wireless/ath/ath12k/accel_cfg.c
+++ b/drivers/net/wireless/ath/ath12k/accel_cfg.c
@@ -67,14 +67,29 @@ static bool ath12k_ds_get_node_id(struct net_device *ndev, const u8 *peer_mac,
 		return false;
 
 	if (ahvif->vdev_type == WMI_VDEV_TYPE_STA) {
-		arvif = ahvif->link[ahvif->primary_link_id];
+		rcu_read_lock();
+		sta = ieee80211_find_sta(vif,vif->cfg.ap_addr);
+
+		if (!sta) {
+			ath12k_warn(NULL, "ieee80211_sta is null");
+			goto unlock_n_fail;
+		}
+
+		ahsta = ath12k_sta_to_ahsta(sta);
+		arvif = (!sta->mlo) ? ahvif->link[ahsta->deflink.link_id] :
+				ahvif->link[ahsta->primary_link_id];
+
+		if (WARN_ON(!arvif))
+			goto unlock_n_fail;
+
 		ab = arvif->ab;
 
 		/* Update and return DS node_id only if the chipset support DS*/
 		if (!test_bit(ATH12K_FLAG_PPE_DS_ENABLED, &ab->dev_flags))
-			return false;
+			goto unlock_n_fail;
 
 		*node_id = ab->ppe.ds_node_id;
+		rcu_read_unlock();
 		return true;
 	}
 
@@ -82,10 +97,8 @@ static bool ath12k_ds_get_node_id(struct net_device *ndev, const u8 *peer_mac,
 		rcu_read_lock();
 
 		sta = ieee80211_find_sta(vif, peer_mac);
-		if (!sta) {
-			rcu_read_unlock();
-			return false;
-		}
+		if (!sta)
+			goto unlock_n_fail;
 
 		ahsta = ath12k_sta_to_ahsta(sta);
 
@@ -94,10 +107,8 @@ static bool ath12k_ds_get_node_id(struct net_device *ndev, const u8 *peer_mac,
 		ab = arvif->ab;
 
 		/* Update and return DS node_id only if the chipset support DS*/
-		if (!test_bit(ATH12K_FLAG_PPE_DS_ENABLED, &ab->dev_flags)) {
-			rcu_read_unlock();
-			return false;
-		}
+		if (!test_bit(ATH12K_FLAG_PPE_DS_ENABLED, &ab->dev_flags))
+			goto unlock_n_fail;
 
 		*node_id = ab->ppe.ds_node_id;
 
@@ -106,6 +117,10 @@ static bool ath12k_ds_get_node_id(struct net_device *ndev, const u8 *peer_mac,
 		return true;
 	}
 	return false;
+
+unlock_n_fail:
+	rcu_read_unlock();
+	return false;
 }
 
 static const struct ath_dp_accel_cfg_ops ath_dp_accel_cfg_ops_obj = {
diff --git a/drivers/net/wireless/ath/ath12k/ppe.c b/drivers/net/wireless/ath/ath12k/ppe.c
index 57f852f..6a563ca 100644
--- a/drivers/net/wireless/ath/ath12k/ppe.c
+++ b/drivers/net/wireless/ath/ath12k/ppe.c
@@ -1504,8 +1504,11 @@ ath12k_dp_rx_ppeds_fse_add_flow_entry(struct ppe_drv_fse_rule_info *ppe_flow_inf
 
 	ahvif = ath12k_vif_to_ahvif(vif);
 
-	/* FST rules are maintained at ag. Always chose default deflink's ab */
-	arvif = (vif->valid_links) ? (ahvif->link[0]) : (&ahvif->deflink);
+	if (!vif->valid_links)
+		arvif = &ahvif->deflink;
+	else
+		arvif = ahvif->link[ffs(vif->valid_links) - 1];
+
 	if (WARN_ON(!arvif))
 		return false;
 
@@ -1548,8 +1551,11 @@ ath12k_dp_rx_ppeds_fse_del_flow_entry(struct ppe_drv_fse_rule_info *ppe_flow_inf
 
 	ahvif = ath12k_vif_to_ahvif(vif);
 
-	/* FST rules are maintained at ag. Always chose default deflink's ab */
-	arvif = (vif->valid_links) ? (ahvif->link[0]) : (&ahvif->deflink);
+	if (!vif->valid_links)
+		arvif = &ahvif->deflink;
+	else
+		arvif = ahvif->link[ffs(vif->valid_links) - 1];
+
 	if (WARN_ON(!arvif))
 		return false;
 
-- 
2.34.1

